### PR TITLE
[DX-2625] port-core-sdk

### PR DESCRIPTION
**Pull request by CodeSchwert**
### Hi👋, please prefix this PR's title with:

- [ ] `breaking-change:` if you have introduced modification that necessitates immediate adjustments by this SDK's users to their applications, clients, or integrations to avert disruptions to existing features or functionalities.
- [x] `feat:`, `fix:`, `refactor:`, `docs:`, or `chore:`.

# Summary

[DX-2526](https://immutable.atlassian.net/browse/DX-2625)

The `x-client` / `IMXClient` class previously depended on the Core SDK for most of it's functionality. This PR ports across all the functionality from the Core SDK into the `x-client` module so users can switch to the Unified SDK / `x-client` module instead of using the soon-to-be-deprecated Core SDK.

The `IMXClient` class should functionality is the same as before, just the internals of it have changed.

# Detail and impact of the change

## Changed
- Core SDK functionality has been ported into the `x-client` module rather than depending on the Core SDK.

# Anything else worth calling out?

### CODE REVIEW

File: packages/x-client/src/types/api.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+/* eslint-disable max-len */
+import { imx } from '@imtbl/generated-clients';
+
+/**
+ * Need to specifically export the classes and interfaces from the generated
+ * clients imx object for rollup to bundle them correctly.
+ */
```
Comment (CodeSchwert) [inline packages/x-client/src/types/api.ts:8]: This file doesn't look ideal, but was required for rollup to bundle the classes and interfaces correctly during build. The alternative was to reexport the `imx` object, but that would have exposed a lot of unused generated code.

File: sdk/package.json [hunk 2]
```diff
@@ -12,7 +12,9 @@
     "@biom3/react": "^0.20.2",
     "@ethersproject/abi": "^5.7.0",
     "@ethersproject/abstract-signer": "^5.7.0",
+    "@ethersproject/keccak256": "^5.7.0",
     "@ethersproject/providers": "^5.7.2",
+    "@ethersproject/strings": "^5.7.0",
     "@ethersproject/wallet": "^5.7.0",
     "@imtbl/core-sdk": "^2.4.0",
```
Comment (carmen0208) [inline sdk/package.json:19]: do we want to remove the dependency?

Comment (CodeSchwert) [reply sdk/package.json:19]: It's used in the `x-client` module.

File: packages/passport/sdk/src/starkEx/workflows/transfer.ts [hunk 3]
```diff
@@ -145,6 +141,7 @@ export async function batchNftTransfer({
       });
 
       const requests = await Promise.all(
+        // @ts-ignore - TS is not able to infer the type of the response
```
Comment (carmen0208) [inline packages/passport/sdk/src/starkEx/workflows/transfer.ts:144]: Why it becomes a problem.
what's the difference between `core-sdk` and `x-client`'s **getSignableTransfer** return?

Comment (CodeSchwert) [reply packages/passport/sdk/src/starkEx/workflows/transfer.ts]: Seems like my VSCode is not complaining about this anymore...

**Review by carmen0208 — APPROVED** [review]

File: packages/x-client/src/types/errors.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { imx } from '@imtbl/generated-clients';
+
+/**
+ * Custom Error class that is returned from the API when a request fails
+ */
+export class IMXError extends Error {
+  /**
+   * [See a list of thrown error codes](https://docs.x.immutable.com/docs/error-codes/)
```
Comment (carleeto) [inline packages/x-client/src/types/errors.ts:8]: This is a broken link.

**Review by carleeto — APPROVED** [review]

File: packages/x-client/src/utils/stark/legacy/crypto/constants.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+const constantPoints = constantPointsHex.map((coords: string[]) =>
+  // eslint-disable-next-line implicit-arrow-linebreak
+  starkEc.curve.point(new BN(coords[0], 16), new BN(coords[1], 16)),
+// eslint-disable-next-line function-paren-newline
+);
+const shiftPoint = constantPoints[0];
+
+// Instruction type mapping encoded in BigNumber
+// see https://docs.starkware.co/starkex-v3/starkex-deep-dive/message-encodings/signatures
```
[the remark this answers is not in the document]
Comment (CodeSchwert) [reply packages/x-client/src/utils/stark/legacy/crypto/constants.ts]: That page doesn't show the instruction mappings to BNs. Removing the link altogether.

**Review by carmen0208 — APPROVED** [review]

**Review by carleeto — APPROVED** [review]

**Pull request merged**